### PR TITLE
Fixed #31340 -- Allowed query expressions in SearchQuery.value and __search lookup.

### PR DESCRIPTION
**Pull request by bmispelon**
ticket-31340

### CODE REVIEW

File: django/contrib/postgres/search.py [hunk 1]
```diff
@@ -178,17 +178,22 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
         if self.config:
             resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if hasattr(self.value, 'resolve_expression'):
+            resolved.value = self.value.resolve_expression(query, allow_joins, reuse, summarize, for_save)
         return resolved
```
Comment (charettes) [inline django/contrib/postgres/search.py]: This makes me wonder why `SearchQuery` is a `Value` in the first place given it needs to resolve `config` and `value` now.

Comment (bmispelon) [reply django/contrib/postgres/search.py]: To be honest I'm a bit lost between `Value`, `Expression`, `Func`, ...
What do you think would be a better base class for this?

Comment (charettes) [reply django/contrib/postgres/search.py]: If look like could subclass `Func` and avoid any `resolve_expression` or `as_sql` overrides.

```python
class SearchQuery(SearchQueryCombinable, Func):
    output_field = SearchQueryField()
    SEARCH_TYPES = {
        'plain': 'plainto_tsquery',
        'phrase': 'phraseto_tsquery',
        'raw': 'to_tsquery',
        'websearch': 'websearch_to_tsquery',
    }
    def __init__(value, output_field=None, *, config=None, invert=False, search_type='plain'):
        function = self.SEARCH_TYPES.get(search_type)
        if function is None:
            raise ValueError("Unknown search_type argument '%s'." % search_type)
        if not hasattr(value, 'resolve_expression'):
           value = Value(value)
        expressions = (value,)
        if config is not None:
            config = SearchConfig.from_parameter(config)
            expressions = (config,) + expressions
       super().__init__(*expressions, output_field=output_field, function=function)
        if invert:
            self.template = '!!(%s)' % Func.template
```

Comment (bmispelon) [reply django/contrib/postgres/search.py]: Not knowing much about the internals, I think changing to `Func` makes sense.

I pushed a new commit that makes `SearchQuery` inherit from `Func` using your proposal but I had to make a few more changes to fix some failing `postgres_tests` tests.

There's still an issue with the tests for `SearchQuery.__repr__` (see commented out tests in the PR) but I don't think it's fixable other than by changing the expected output of the tests to the new one.

I'm also concerned about potential backwards compatibility issues. I'm not sure how much of the API of `SearchQuery` is considered public. It shouldn't be too hard to add deprecated properties for attributes like `config`, `search_type`, `invert`, ...
Do you think it's worth the effort?

File: django/contrib/postgres/search.py [hunk 2]
```diff
@@ -167,50 +167,35 @@ class SearchQuery(SearchQueryCombinable, Value):
...
-        combined = super()._combine(other, connector, reversed)
-        combined.output_field = SearchQueryField()
-        return combined
+        if not hasattr(value, 'resolve_expression'):
+           value = Value(value)
+        expressions = (value,)
+        if config is not None:
+            config = SearchConfig.from_parameter(config)
+            expressions = (config,) + expressions
+        super().__init__(*expressions, output_field=output_field, function=function, invert=invert)
```
Comment (felixxm) [inline django/contrib/postgres/search.py]: We can keep `invert` directly in the `SearchQuery` instead of `extra`, e.g.
```python
...
self.invert = invert
super().__init__(*expressions, output_field=output_field, function=function)
```

File: django/contrib/postgres/search.py [hunk 3]
```diff
@@ -167,50 +167,35 @@ class SearchQuery(SearchQueryCombinable, Value):
...
         result = super().__str__()
-        return ('~%s' % result) if self.invert else result
+        return ('~%s' % result) if self.extra['invert'] else result
 
+    @property
+    def template(self):
+        tpl = super().template
+        if self.extra['invert']:
+            tpl = '!!({})'.format(tpl)
+        return tpl
```
Comment (felixxm) [inline django/contrib/postgres/search.py]: I would handle this in `as_sql()`, i.e.
```python
def as_sql(self, compiler, connection, template=None, **extra_context):
    sql, params = super().as_sql(compiler, connection, template, **extra_context)
    if self.invert:
        sql = '!!({})'.format(sql)
    return sql, params
```

Comment (bmispelon) [reply django/contrib/postgres/search.py]: That seems like a cleaner option than my `property` approach. I've pushed a commit with that, thanks.

**Comment by felixxm** [comment]
I moved all cleanups unrelated with a bugfix to a separate PR #12563.

File: django/contrib/postgres/search.py [hunk 4]
```diff
@@ -167,50 +167,40 @@ class SearchQuery(SearchQueryCombinable, Value):
...
+        super().__init__(*expressions, output_field=output_field, function=function)
 
-    def _combine(self, other, connector, reversed):
-        combined = super()._combine(other, connector, reversed)
-        combined.output_field = SearchQueryField()
-        return combined
+    def copy(self):
+        copy = super().copy()
+        copy.invert = self.invert
+        return copy
```
Comment (felixxm) [inline django/contrib/postgres/search.py]: `copy()` in unnecessary.

File: django/contrib/postgres/search.py [hunk 5]
```diff
@@ -167,50 +167,40 @@ class SearchQuery(SearchQueryCombinable, Value):
...
+        if config is not None:
+            config = SearchConfig.from_parameter(config)
+            expressions = (config,) + expressions
+        self.invert = invert
+        super().__init__(*expressions, output_field=output_field, function=function)
 
-    def _combine(self, other, connector, reversed):
-        combined = super()._combine(other, connector, reversed)
-        combined.output_field = SearchQueryField()
-        return combined
```
Comment (felixxm) [inline django/contrib/postgres/search.py]: `_combine()` is unnecessary since its introduction, I moved it to a separate commit b62c58d5fccaeedc7bc4a064b4678e4ed1e77581.

File: django/contrib/postgres/search.py [hunk 6]
```diff
@@ -138,8 +138,8 @@ def _combine(self, other, connector, reversed):
                 'instances, got %s.' % type(other).__name__
             )
         if reversed:
-            return CombinedSearchQuery(other, connector, self, self.config)
-        return CombinedSearchQuery(self, connector, other, self.config)
```
Comment (felixxm) [inline django/contrib/postgres/search.py]: Combined `SearchQuery`'s should have configs, I added a new test in e2417010daf697d5a58bc2057a24b7785c4cef03 to prevent future regressions.

**Comment by felixxm** [comment]
I rebased from master (after #12563). I'm going to push minor edits in the next few days.

**Comment by felixxm** [comment]
@bmispelon Thanks again :+1: I added release notes and simplified tests.
